File: joptimize/test/src/examples/opt/InterfacePreservation.scala
package joptimize.examples.opt

object InterfacePreservation {
  def shallow(i: Int) = new Bar().run(i)
  def deep(i: Int) = new Qux().run(i)

}


trait Foo
class Bar() extends Foo{
  def run(i: Int) = i + 1
}
trait Baz extends Foo
class Qux() extends Baz{
  def run(i: Int) = i + 2
}

File: joptimize/test/src/examples/opt/InterfacePreservation.java
package joptimize.examples.opt;

class InterfacePreservation {
    static int shallow(int i) { return new Bar().run(i); }
    static int deep(int i) {return new Qux().run(i); }

    interface Foo{}
    static class Bar implements Foo{
        int run(int i){ return i + 1; }
    }
    interface Baz extends Foo{}
    static class Qux implements Baz{
        int run(int i){ return i + 2; }
    }
}

